pkg/helm/release/manager_factory.go

```diff
@@ -113,8 +122,66 @@ func getReleaseServer(cr *unstructured.Unstructured, storageBackend *storage.Sto
 	return tiller.NewReleaseServer(env, cs, false), nil
 }
 
-func getReleaseName(cr *unstructured.Unstructured) string {
-	return fmt.Sprintf("%s-%s", cr.GetName(), shortenUID(cr.GetUID()))
+// getReleaseName returns a release name for the CR. If a release for the
+// legacy name exists, the legacy name is returned. This ensures
+// backwards-compatibility for pre-existing CRs.
+//
+// If no releases are found with the legacy name, getReleaseName searches for
+// a release using the CR name. If a release cannot be found, or if it is found
+// and was created by the chart managed by this manager, the CR name is
+// returned.
+//
+// If a release is found but it was created by another chart, that means we
+// have a release name collision, so return an error. This case is possible
+// because Kubernetes allows instances of different types to have the same name
+// in the same namespace.
+//
+//     NOTE: The motivation for including the CR's UID was to prevent any
+//     possibility of a collision between release names of CRs of different
+//     types, so we now have to take extra precautions.
+//
+// The reason for this change is based on an interaction between the Kubernetes
+// constraint that limits label values to 63 characters and the Helm convention
+// of including the release name as a label on release resources.
+//
+// Since the legacy release name includes a 25-character value based on the
+// parent CR's UID, it leaves little extra space for the CR name and any other
+// identifying names or characters added by templates.
+//
+// TODO(jlanford): As noted above, using the CR name as the release name raises
+// the possibility of collision. We should move this logic to a validating
+// admission webhook so that the CR owner receives immediate feedback of the
+// collision. As is, the only indication of collision will be in the CR status
+// and operator logs.
+func getReleaseName(storageBackend *storage.Storage, crChartName string, cr *unstructured.Unstructured) (string, error) {
+	// If a release with the legacy name exists, return the legacy name.
+	legacyName := fmt.Sprintf("%s-%s", cr.GetName(), shortenUID(cr.GetUID()))
+	_, err := storageBackend.History(legacyName)
+	if err == nil {
+		return legacyName, nil
+	} else if !notFoundErr(err) {
+		return "", err
+	}
+
+	// If a release with the CR name does not exist, return the CR name.
+	releaseName := cr.GetName()
+	history, err := storageBackend.History(releaseName)
+	if notFoundErr(err) || len(history) == 0 {
```

Comment:
```suggestion
	if notFoundErr(err) || (err == nil && len(history) == 0) {
```

in case `notFoundErr(err)` is false when `err != nil` (we want to return this error) and `len(history) == 0` is true.